preview/MsixCore/msixmgr/CryptoProvider.cpp

```diff
@@ -0,0 +1,183 @@
+            {
+                return;
+            }
+            this->hashHandle = NULL;
+        }
+
+        if (NULL != this->providerHandle)
+        {
+            if (!SUCCEEDED(BcryptLibrary::BCryptCloseAlgorithmProvider(this->providerHandle, 0)))
+            {
+                return;
+            }
+            this->providerHandle = NULL;
+        }
+    }
+
+    HRESULT CryptoProvider::Create(
+        _Outptr_ CryptoProvider** provider)
+    {
+        RETURN_IF_FAILED(BcryptLibrary::Load());
+
+        std::unique_ptr<CryptoProvider> cp(new APPXCOMMON_NEW_TAG CryptoProvider());
+        if (cp == nullptr)
+        {
+            return HRESULT_FROM_WIN32(ERROR_OUTOFMEMORY);
+        }
+
+        *provider = cp.release();
+        return S_OK;
+    }
+
+    HRESULT CryptoProvider::OpenProvider()
+    {
+        PCWSTR algorithmId = BCRYPT_SHA256_ALGORITHM;
+        PCWSTR implementation = NULL;
+        ULONG flags = 0;
+
+        BcryptLibrary::BCryptOpenAlgorithmProvider(
+            &this->providerHandle,
+            algorithmId,
+            implementation,
+            flags);
+
+        return S_OK;
+    }
+
+    HRESULT CryptoProvider::StartDigest()
+    {
+        BYTE* hashObject;
+        ULONG hashObjectSize;
+        ULONG resultSize;
+
+        RETURN_IF_FAILED(OpenProvider());
+
+        BcryptLibrary::BCryptGetProperty(
+            this->providerHandle,
+            BCRYPT_OBJECT_LENGTH,
+            (PUCHAR)&hashObjectSize,
+            sizeof(hashObjectSize),
+            &resultSize,
+            0);
+
+        if (sizeof(this->quickHashObjectBuffer) >= hashObjectSize)
+        {
+            hashObject = this->quickHashObjectBuffer;
+        }
+        else
+        {
+            RETURN_IF_FAILED(this->hashObjectBuffer.SetLength(hashObjectSize));
+            hashObject = this->hashObjectBuffer.GetBufferPtr();
+        }
+
+        BcryptLibrary::BCryptCreateHash(
+            this->providerHandle,
+            &this->hashHandle,
+            hashObject,
+            hashObjectSize,
+            NULL,
+            NULL,
+            0);
+
+        return S_OK;
+    }
+
+    HRESULT CryptoProvider::DigestData(
+        _In_ const COMMON_BYTES* data)
+    {
+        BcryptLibrary::BCryptHashData(
+            this->hashHandle,
+            data->bytes,
+            data->length,
+            0);
+
+        return S_OK;
+    }
+
+    HRESULT CryptoProvider::GetDigest(
+        _Out_ COMMON_BYTES* digest)
+    {
+        BYTE* digestPtr;
+        ULONG digestSize;
+
+        if (0 == this->digest.length)
+        {
+            NTSTATUS status;
+
+            ULONG resultSize;
+
+            status = BcryptLibrary::BCryptGetProperty(
+                this->providerHandle,
+                BCRYPT_HASH_LENGTH,
+                (PUCHAR)&digestSize,
+                sizeof(digestSize),
+                &resultSize,
+                0);
+
+            if (FAILED(status))

```

Comment:
>FAILED [](start = 16, length = 6)

FAILED is for HRESULTS. the return here is NTSTATUS, you should convert status to HRESULT first (i.e. FAILED(HRESULT_FROM_NT(status))

Reply:
Replaced with HRESULT_FROM_NT.